src/fhirtypes/ElementDefinition.ts

```diff
@@ -22,6 +22,67 @@ import {
 } from '../errors';
 import { setPropertyOnInstance } from './common';
 
+export class ElementDefinitionType {
+  private _code: string;
+  profile?: string[];
+  targetProfile?: string[];
+  aggregation?: string[];
+  versioning?: string;
+  extension?: ElementDefinitionExtension[];
+
+  constructor(code?: string) {
+    this._code = code;
+  }
+
+  /**
+   * Element.id type is specified in the valueUrl of an extension.
+   * This function returns the extension valueUrl if available, else returns the code.
+   * @see {@link http://hl7.org/fhir/extension-structuredefinition-fhir-type.html}
+   */
+  get code(): string {
+    if (this.extension) {
+      const fhirTypeExtension = this.extension.find(
+        ext => ext.url === 'http://hl7.org/fhir/StructureDefinition/structuredefinition-fhir-type'
+      );
+      if (fhirTypeExtension) return fhirTypeExtension.valueUrl;
+      return this._code;
+    }
+    return this._code;
+  }
+
+  set code(c: string) {
+    this._code = c;
+  }
+
+  getActualCode(): string {
+    return this._code;
+  }
+
+  toJSON(): ElementDefinitionTypeJSON {
+    // Remove the _code key specific to ElementDefinitionType
+    const elDefTypeClone = cloneDeep(this);
+    delete elDefTypeClone._code;
+
+    // Create ElementDefinitionTypeJSON with a code and any properties present on the ElementDefinitionType
+    const elDefTypeJSON: ElementDefinitionTypeJSON = {
+      code: this.getActualCode(),
+      ...elDefTypeClone
+    };
+    return elDefTypeJSON;
+  }
+
+  static fromJSON(json: any): ElementDefinitionType {
+    const elDefType = new ElementDefinitionType();
+    elDefType._code = json.code;
+    if (json.profile) elDefType.profile = json.profile;
+    if (json.targetProfile) elDefType.targetProfile = json.targetProfile;
+    if (json.aggregation) elDefType.aggregation = json.aggregation;
+    if (json.versioning) elDefType.versioning = json.versioning;
+    if (json.extension) elDefType.extension = json.extension;
```

Comment:
Hm this is a good point. I could have all the properties set, but with various `undefined`s, or I could only assign the properties that are defined on the type `json` by doing something along these lines (I think cleaner than the if's):

```
const jsonClone = cloneDeep(json);
delete jsonClone.code;
elDefType = { ...elDefType, ...jsonClone };
```
Do you have a preference? Is there a reason we'd want the `undefined`s?

Reply:
Since `elDefType` is a class instance now, I don't think the object destructuring would work, would it?

I think setting something that isn't there as `undefined` is kind of like a no-op.  I don't think it does any harm.  I don't think.  But I hear you re: being a bit more precise about explicitly setting only what is in the JSON.  I guess another approach would be something like:
```ts
Object.keys(json).forEach(key => elDefType[key] = json[key];
```
TypeScript might yell at you though.  In the end, the `if` statements aren't a super big deal if that's your preference.  I just wasn't sure they were necessary is all.

Reply:
I believe you are correct. The `undefined`s don't do any harm, as far as I can see. I do think typescript will be upset by setting keys in that way. I'm fine with just setting each property directly - essentially the same as what is currently there, but without the if statements. As long as that is what you were originally suggesting and I'm not misunderstanding.

Reply:
Yep!  Sounds good!